File: client/python/src/at_runner/session.py
"""AT Runner client library — wraps generated gRPC stubs into a clean Python API."""

from __future__ import annotations

import dataclasses
from pathlib import Path
from typing import Callable, Optional, Sequence

import grpc

from at_runner._generated.at.runner.v1 import runner_pb2, runner_pb2_grpc

_MAX_MSG = 256 * 1024 * 1024  # 256 MiB
_DEFAULT_CHANNEL_OPTIONS = [
    ("grpc.max_send_message_length", _MAX_MSG),
    ("grpc.max_receive_message_length", _MAX_MSG),
]


# ---------------------------------------------------------------------------
# Result dataclasses
# ---------------------------------------------------------------------------

@dataclasses.dataclass
class RunResult:
    status: str
    exit_code: int
    stdout: str
    stderr: str
    elapsed: float
    files: dict[str, bytes]


@dataclasses.dataclass
class StepResult:
    status: str
    exit_code: int
    elapsed: float
    files: dict[str, bytes]


@dataclasses.dataclass
class PipelineResult:
    all_succeeded: bool
    elapsed: float
    steps: dict[str, StepResult]
    skipped_steps: list[str]


@dataclasses.dataclass
class Step:
    id: str
    model: str
    file_root: str
    inputs: dict[str, bytes] | None = None
    depends_on: list[str] | None = None
    timeout: int | None = None


# ---------------------------------------------------------------------------
# Status enum mapping
# ---------------------------------------------------------------------------

_STATUS_MAP = {
    runner_pb2.RUN_STATUS_UNSPECIFIED: "unspecified",
    runner_pb2.RUN_STATUS_COMPLETED: "completed",
    runner_pb2.RUN_STATUS_TIMED_OUT: "timed_out",
    runner_pb2.RUN_STATUS_SIGNALED: "signaled",
    runner_pb2.RUN_STATUS_SKIPPED: "skipped",
}


def _status_str(code: int) -> str:
    return _STATUS_MAP.get(code, f"unknown({code})")


# ---------------------------------------------------------------------------
# Tier 1: run_sync (standalone function)
# ---------------------------------------------------------------------------

def run_sync(
    target: str,
    model: str,
    file_root: str,
    inputs: dict[str, bytes] | None = None,
    timeout: int | None = None,
    *,
    channel_options: list | None = None,
) -> RunResult:
    """One-shot run: send files, get files back. No session required."""
    opts = list(channel_options or []) + _DEFAULT_CHANNEL_OPTIONS
    with grpc.insecure_channel(target, options=opts) as channel:
        stub = runner_pb2_grpc.RunnerStub(channel)
        req = runner_pb2.RunSyncRequest(
            model=model,
            file_root=file_root,
            inputs=[
                runner_pb2.File(name=n, content=_to_bytes(c))
                for n, c in (inputs or {}).items()
            ],
        )
        if timeout is not None:
            req.timeout_seconds = timeout

        resp = stub.RunSync(req)

    return RunResult(
        status=_status_str(resp.status),
        exit_code=resp.exit_code,
        stdout=resp.stdout.decode("utf-8", errors="replace"),
        stderr=resp.stderr.decode("utf-8", errors="replace"),
        elapsed=resp.elapsed_seconds,
        files={f.name: f.content for f in resp.outputs},
    )


# ---------------------------------------------------------------------------
# Tier 2 + 3: ATSession
# ---------------------------------------------------------------------------

class ATSession:
    """Interactive session backed by a persistent gRPC connection."""

    def __init__(self, target: str, *, channel_options: list | None = None):
        opts = list(channel_options or []) + _DEFAULT_CHANNEL_OPTIONS
        self._channel = grpc.insecure_channel(target, options=opts)
        self._stub = runner_pb2_grpc.RunnerStub(self._channel)

    def close(self) -> None:
        self._channel.close()

    def __enter__(self):
        return self

    def __exit__(self, *exc):
        self.close()

    # --- Workspace management ---

    def upload(self, name: str, content: bytes | str | Path) -> tuple[str, int]:
        """Upload a file. Returns (name, size_bytes)."""
        data = _to_bytes(content)

        def _chunks():
            chunk_size = 65536
            yield runner_pb2.FileChunk(name=name, data=data[:chunk_size])
            offset = chunk_size
            while offset < len(data):
                end = min(offset + chunk_size, len(data))
                yield runner_pb2.FileChunk(data=data[offset:end])
                offset = end

        resp = self._stub.UploadFile(_chunks())
        return resp.name, resp.size_bytes

    def download(self, name: str) -> bytes:
        """Download a file from the workspace."""
        stream = self._stub.GetFile(runner_pb2.GetFileRequest(name=name))
        buf = bytearray()
        for chunk in stream:
            buf.extend(chunk.data)
        return bytes(buf)

    def delete(self, name: str) -> None:
        self._stub.DeleteFile(runner_pb2.DeleteFileRequest(name=name))

    def list_files(self) -> list[tuple[str, int]]:
        """Returns list of (name, size_bytes)."""
        resp = self._stub.ListFiles(runner_pb2.ListFilesRequest())
        return [(f.name, f.size_bytes) for f in resp.files]

    # --- Tier 2: Run (streaming) ---

    def run(
        self,
        model: str,
        file_root: str,
        on_output: Callable[[str, bytes], None] | None = None,
        timeout: int | None = None,
    ) -> RunResult:
        """Run a model in the workspace. Blocks until complete."""
        req = runner_pb2.RunRequest(model=model, file_root=file_root)
        if timeout is not None:
            req.timeout_seconds = timeout

        stream = self._stub.Run(req)
        return _consume_run_stream(stream, on_output)

    # --- Tier 3: RunPipeline ---

    def run_pipeline(
        self,
        steps: Sequence[Step],
        on_step_output: Callable[[str, str, bytes], None] | None = None,
        timeout: int | None = None,
    ) -> PipelineResult:
        """Submit a DAG pipeline. Blocks until complete."""
        pb_steps = []
        for s in steps:
            ps = runner_pb2.PipelineStep(
                id=s.id,
                model=s.model,
                file_root=s.file_root,
                inputs=[
                    runner_pb2.File(name=n, content=_to_bytes(c))
                    for n, c in (s.inputs or {}).items()
                ],
                depends_on=s.depends_on or [],
            )
            if s.timeout is not None:
                ps.timeout_seconds = s.timeout
            pb_steps.append(ps)

        req = runner_pb2.RunPipelineRequest(steps=pb_steps)
        if timeout is not None:
            req.timeout_seconds = timeout

        stream = self._stub.RunPipeline(req)
        return _consume_pipeline_stream(stream, on_step_output)


# ---------------------------------------------------------------------------
# Internal stream consumers
# ---------------------------------------------------------------------------

def _consume_run_stream(
    stream,
    on_output: Callable[[str, bytes], None] | None,
) -> RunResult:
    stdout_parts: list[bytes] = []
    stderr_parts: list[bytes] = []
    status = "unspecified"
    exit_code = -1
    elapsed = 0.0
    files: dict[str, bytes] = {}
    current_file: str | None = None
    current_data: list[bytes] = []

    for msg in stream:
        which = msg.WhichOneof("payload")
        if which == "started":
            pass
        elif which == "output":
            chunk = msg.output
            stream_name = "stdout" if chunk.stream == runner_pb2.OUTPUT_STREAM_STDOUT else "stderr"
            if stream_name == "stdout":
                stdout_parts.append(chunk.data)
            else:
                stderr_parts.append(chunk.data)
            if on_output:
                on_output(stream_name, chunk.data)
        elif which == "completed":
            status = _status_str(msg.completed.status)
            exit_code = msg.completed.exit_code
            elapsed = msg.completed.elapsed_seconds
        elif which == "file":
            fc = msg.file
            if fc.name:
                if current_file is not None:
                    files[current_file] = b"".join(current_data)
                current_file = fc.name
                current_data = [fc.data]
            else:
                current_data.append(fc.data)

    if current_file is not None:
        files[current_file] = b"".join(current_data)

    return RunResult(
        status=status,
        exit_code=exit_code,
        stdout=b"".join(stdout_parts).decode("utf-8", errors="replace"),
        stderr=b"".join(stderr_parts).decode("utf-8", errors="replace"),
        elapsed=elapsed,
        files=files,
    )


def _consume_pipeline_stream(
    stream,
    on_step_output: Callable[[str, str, bytes], None] | None,
) -> PipelineResult:
    step_results: dict[str, dict] = {}
    all_succeeded = False
    elapsed = 0.0
    skipped_steps: list[str] = []

    # Track current file per step
    current_files: dict[str, tuple[str, list[bytes]]] = {}

    def _finalize_file(step_id: str) -> None:
        if step_id in current_files:
            fname, parts = current_files.pop(step_id)
            sr = step_results.setdefault(step_id, {"files": {}})
            sr["files"][fname] = b"".join(parts)

    for msg in stream:
        which = msg.WhichOneof("payload")
        if which == "pipeline_started":
            pass
        elif which == "step":
            ev = msg.step
            sid = ev.step_id
            detail = ev.WhichOneof("detail")
            if detail == "started":
                step_results.setdefault(sid, {"files": {}})
            elif detail == "output":
                chunk = ev.output
                stream_name = "stdout" if chunk.stream == runner_pb2.OUTPUT_STREAM_STDOUT else "stderr"
                if on_step_output:
                    on_step_output(sid, stream_name, chunk.data)
            elif detail == "completed":
                _finalize_file(sid)
                sr = step_results.setdefault(sid, {"files": {}})
                sr["status"] = _status_str(ev.completed.status)
                sr["exit_code"] = ev.completed.exit_code
                sr["elapsed"] = ev.completed.elapsed_seconds
            elif detail == "file":
                fc = ev.file
                if fc.name:
                    _finalize_file(sid)
                    current_files[sid] = (fc.name, [fc.data])
                else:
                    if sid in current_files:
                        current_files[sid][1].append(fc.data)
        elif which == "pipeline_completed":
            # Finalize any remaining files
            for sid in list(current_files):
                _finalize_file(sid)
            all_succeeded = msg.pipeline_completed.all_succeeded
            elapsed = msg.pipeline_completed.elapsed_seconds
            skipped_steps = list(msg.pipeline_completed.skipped_steps)

    steps = {}
    for sid, data in step_results.items():
        steps[sid] = StepResult(
            status=data.get("status", "unspecified"),
            exit_code=data.get("exit_code", -1),
            elapsed=data.get("elapsed", 0.0),
            files=data.get("files", {}),
        )

    return PipelineResult(
        all_succeeded=all_succeeded,
        elapsed=elapsed,
        steps=steps,
        skipped_steps=skipped_steps,
    )


# ---------------------------------------------------------------------------
# Utilities
# ---------------------------------------------------------------------------

def _to_bytes(content: bytes | str | Path) -> bytes:
    if isinstance(content, bytes):
        return content
    if isinstance(content, str):
        return content.encode("utf-8")
    if isinstance(content, Path):
        return content.read_bytes()
    raise TypeError(f"expected bytes, str, or Path; got {type(content)}")

File: client/rust/src/session.rs
use std::collections::HashMap;

use tonic::transport::Channel;

use crate::proto;
use crate::proto::runner_client::RunnerClient;

#[derive(Debug, Clone)]
pub struct RunResult {
    pub status: String,
    pub exit_code: i32,
    pub stdout: String,
    pub stderr: String,
    pub elapsed: f64,
    pub files: HashMap<String, Vec<u8>>,
}

#[derive(Debug, Clone)]
pub struct StepResult {
    pub status: String,
    pub exit_code: i32,
    pub elapsed: f64,
    pub files: HashMap<String, Vec<u8>>,
}

#[derive(Debug, Clone)]
pub struct PipelineResult {
    pub all_succeeded: bool,
    pub elapsed: f64,
    pub steps: HashMap<String, StepResult>,
    pub skipped_steps: Vec<String>,
}

#[derive(Debug, Clone)]
pub struct Step {
    pub id: String,
    pub model: String,
    pub file_root: String,
    pub inputs: Vec<(String, Vec<u8>)>,
    pub depends_on: Vec<String>,
    pub timeout: Option<u32>,
}

impl Step {
    pub fn new(id: &str, model: &str, file_root: &str) -> Self {
        Self {
            id: id.to_string(),
            model: model.to_string(),
            file_root: file_root.to_string(),
            inputs: Vec::new(),
            depends_on: Vec::new(),
            timeout: None,
        }
    }

    pub fn with_input(mut self, name: &str, data: &[u8]) -> Self {
        self.inputs.push((name.to_string(), data.to_vec()));
        self
    }

    pub fn depends_on(mut self, deps: &[&str]) -> Self {
        self.depends_on = deps.iter().map(|s| s.to_string()).collect();
        self
    }
}

pub struct ATSession {
    client: RunnerClient<Channel>,
    chunk_size: usize,
}

impl ATSession {
    pub async fn connect(target: &str) -> Result<Self, Box<dyn std::error::Error>> {
        let client = RunnerClient::connect(format!("http://{target}")).await?;
        Ok(Self {
            client,
            chunk_size: 65536,
        })
    }

    pub async fn upload(
        &mut self,
        name: &str,
        data: &[u8],
    ) -> Result<(String, u64), Box<dyn std::error::Error>> {
        let chunks = build_upload_chunks(name, data, self.chunk_size);
        let resp = self
            .client
            .upload_file(tokio_stream::iter(chunks))
            .await?
            .into_inner();
        Ok((resp.name, resp.size_bytes))
    }

    pub async fn download(&mut self, name: &str) -> Result<Vec<u8>, Box<dyn std::error::Error>> {
        let resp = self
            .client
            .get_file(proto::GetFileRequest {
                name: name.to_string(),
            })
            .await?;

        let mut stream = resp.into_inner();
        let mut buf = Vec::new();
        while let Some(chunk) = stream.message().await? {
            buf.extend_from_slice(&chunk.data);
        }
        Ok(buf)
    }

    pub async fn delete(&mut self, name: &str) -> Result<(), Box<dyn std::error::Error>> {
        self.client
            .delete_file(proto::DeleteFileRequest {
                name: name.to_string(),
            })
            .await?;
        Ok(())
    }

    pub async fn list_files(&mut self) -> Result<Vec<(String, u64)>, Box<dyn std::error::Error>> {
        let resp = self
            .client
            .list_files(proto::ListFilesRequest {})
            .await?
            .into_inner();
        Ok(resp
            .files
            .into_iter()
            .map(|f| (f.name, f.size_bytes))
            .collect())
    }

    pub async fn run(
        &mut self,
        model: &str,
        file_root: &str,
    ) -> Result<RunResult, Box<dyn std::error::Error>> {
        self.run_with_timeout(model, file_root, None).await
    }

    pub async fn run_with_timeout(
        &mut self,
        model: &str,
        file_root: &str,
        timeout: Option<u32>,
    ) -> Result<RunResult, Box<dyn std::error::Error>> {
        let req = proto::RunRequest {
            model: model.to_string(),
            file_root: file_root.to_string(),
            timeout_seconds: timeout,
        };

        let mut stream = self.client.run(req).await?.into_inner();
        let mut stdout_parts: Vec<Vec<u8>> = Vec::new();
        let mut stderr_parts: Vec<Vec<u8>> = Vec::new();
        let mut status = "unspecified".to_string();
        let mut exit_code = -1i32;
        let mut elapsed = 0.0f64;
        let mut files: HashMap<String, Vec<u8>> = HashMap::new();
        let mut current_file: Option<String> = None;
        let mut current_data: Vec<Vec<u8>> = Vec::new();

        while let Some(msg) = stream.message().await? {
            if let Some(payload) = msg.payload {
                match payload {
                    proto::run_output::Payload::Started(_) => {}
                    proto::run_output::Payload::Output(chunk) => {
                        if chunk.stream == proto::OutputStream::Stdout as i32 {
                            stdout_parts.push(chunk.data);
                        } else {
                            stderr_parts.push(chunk.data);
                        }
                    }
                    proto::run_output::Payload::Completed(c) => {
                        status = crate::status_str(c.status);
                        exit_code = c.exit_code;
                        elapsed = c.elapsed_seconds;
                    }
                    proto::run_output::Payload::File(fc) => {
                        if !fc.name.is_empty() {
                            if let Some(name) = current_file.take() {
                                files.insert(name, current_data.concat());
                                current_data.clear();
                            }
                            current_file = Some(fc.name);
                        }
                        current_data.push(fc.data);
                    }
                }
            }
        }

        if let Some(name) = current_file {
            files.insert(name, current_data.concat());
        }

        Ok(RunResult {
            status,
            exit_code,
            stdout: String::from_utf8_lossy(&stdout_parts.concat()).to_string(),
            stderr: String::from_utf8_lossy(&stderr_parts.concat()).to_string(),
            elapsed,
            files,
        })
    }

    pub async fn run_pipeline(
        &mut self,
        steps: &[Step],
    ) -> Result<PipelineResult, Box<dyn std::error::Error>> {
        self.run_pipeline_with_timeout(steps, None).await
    }

    pub async fn run_pipeline_with_timeout(
        &mut self,
        steps: &[Step],
        timeout: Option<u32>,
    ) -> Result<PipelineResult, Box<dyn std::error::Error>> {
        let pb_steps: Vec<proto::PipelineStep> = steps
            .iter()
            .map(|s| proto::PipelineStep {
                id: s.id.clone(),
                model: s.model.clone(),
                file_root: s.file_root.clone(),
                inputs: s
                    .inputs
                    .iter()
                    .map(|(name, data)| proto::File {
                        name: name.clone(),
                        content: data.clone(),
                    })
                    .collect(),
                depends_on: s.depends_on.clone(),
                timeout_seconds: s.timeout,
            })
            .collect();

        let req = proto::RunPipelineRequest {
            steps: pb_steps,
            timeout_seconds: timeout,
        };

        let mut stream = self.client.run_pipeline(req).await?.into_inner();

        let mut step_results: HashMap<String, StepResult> = HashMap::new();
        let mut current_files: HashMap<String, (String, Vec<Vec<u8>>)> = HashMap::new();
        let mut all_succeeded = false;
        let mut elapsed = 0.0f64;
        let mut skipped_steps = Vec::new();

        while let Some(msg) = stream.message().await? {
            if let Some(payload) = msg.payload {
                match payload {
                    proto::pipeline_output::Payload::PipelineStarted(_) => {}
                    proto::pipeline_output::Payload::Step(ev) => {
                        let sid = ev.step_id.clone();
                        if let Some(detail) = ev.detail {
                            match detail {
                                proto::step_event::Detail::Started(_) => {
                                    step_results.entry(sid).or_insert_with(|| StepResult {
                                        status: "unspecified".to_string(),
                                        exit_code: -1,
                                        elapsed: 0.0,
                                        files: HashMap::new(),
                                    });
                                }
                                proto::step_event::Detail::Output(_) => {}
                                proto::step_event::Detail::Completed(c) => {
                                    finalize_file(&sid, &mut current_files, &mut step_results);
                                    let sr =
                                        step_results.entry(sid).or_insert_with(|| StepResult {
                                            status: "unspecified".to_string(),
                                            exit_code: -1,
                                            elapsed: 0.0,
                                            files: HashMap::new(),
                                        });
                                    sr.status = crate::status_str(c.status);
                                    sr.exit_code = c.exit_code;
                                    sr.elapsed = c.elapsed_seconds;
                                }
                                proto::step_event::Detail::File(fc) => {
                                    if !fc.name.is_empty() {
                                        finalize_file(&sid, &mut current_files, &mut step_results);
                                        current_files.insert(sid, (fc.name, vec![fc.data]));
                                    } else if let Some((_, parts)) = current_files.get_mut(&sid) {
                                        parts.push(fc.data);
                                    }
                                }
                            }
                        }
                    }
                    proto::pipeline_output::Payload::PipelineCompleted(c) => {
                        for sid in current_files.keys().cloned().collect::<Vec<_>>() {
                            finalize_file(&sid, &mut current_files, &mut step_results);
                        }
                        all_succeeded = c.all_succeeded;
                        elapsed = c.elapsed_seconds;
                        skipped_steps = c.skipped_steps;
                    }
                }
            }
        }

        Ok(PipelineResult {
            all_succeeded,
            elapsed,
            steps: step_results,
            skipped_steps,
        })
    }
}

fn build_upload_chunks(name: &str, data: &[u8], chunk_size: usize) -> Vec<proto::FileChunk> {
    let mut chunks = Vec::new();
    let mut offset = 0;
    let mut first = true;
    while offset < data.len() || first {
        let end = (offset + chunk_size).min(data.len());
        chunks.push(proto::FileChunk {
            name: if first {
                name.to_string()
            } else {
                String::new()
            },
            data: data[offset..end].to_vec(),
        });
        first = false;
        offset = end;
    }
    chunks
}

fn finalize_file(
    sid: &str,
    current_files: &mut HashMap<String, (String, Vec<Vec<u8>>)>,
    step_results: &mut HashMap<String, StepResult>,
) {
    if let Some((fname, parts)) = current_files.remove(sid) {
        let sr = step_results
            .entry(sid.to_string())
            .or_insert_with(|| StepResult {
                status: "unspecified".to_string(),
                exit_code: -1,
                elapsed: 0.0,
                files: HashMap::new(),
            });
        sr.files.insert(fname, parts.concat());
    }
}
